File: android-sms-reader/app/src/main/java/com/rjworldbd/smsreader/ui/MainActivity.java
package com.rjworldbd.smsreader.ui;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;
import com.rjworldbd.smsreader.R;
import com.rjworldbd.smsreader.data.PaymentDbHelper;
import com.rjworldbd.smsreader.service.SmsReaderForegroundService;
import com.rjworldbd.smsreader.sync.SyncManager;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends Activity {

    private static final int PERMISSION_REQUEST_CODE = 1001;
    private TextView tvStatus;
    private TextView tvDiagnosticLogs;
    private Button btnManualSync;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvStatus = findViewById(R.id.tv_service_status);
        tvDiagnosticLogs = findViewById(R.id.tv_diagnostic_logs);
        btnManualSync = findViewById(R.id.btn_manual_sync);

        if (btnManualSync != null) {
            btnManualSync.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Toast.makeText(MainActivity.this, "Triggering sync to server...", Toast.LENGTH_SHORT).show();
                    SyncManager.triggerSync(MainActivity.this);
                }
            });
        }

        SyncManager.setLogListener(new SyncManager.SyncLogListener() {
            @Override
            public void onLogUpdated(final String fullLogText) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        if (tvDiagnosticLogs != null) {
                            tvDiagnosticLogs.setText(fullLogText);
                        }
                        updateStatusDisplay();
                    }
                });
            }
        });

        checkAndRequestPermissions();
        updateStatusDisplay();

        // Trigger an immediate sync attempt on app launch
        SyncManager.triggerSync(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateStatusDisplay();
        if (tvDiagnosticLogs != null) {
            tvDiagnosticLogs.setText(SyncManager.getDiagnosticLogsText());
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        SyncManager.setLogListener(null);
    }

    private void updateStatusDisplay() {
        if (tvStatus == null) return;
        PaymentDbHelper db = PaymentDbHelper.getInstance(this);
        int total = db.getTotalCount();
        int pending = db.getPendingCount();
        int synced = db.getSyncedCount();
        tvStatus.setText("Monitoring Active\nTotal: " + total + " | Pending: " + pending + " | Synced: " + synced);
    }

    private void checkAndRequestPermissions() {
        List<String> permissions = new ArrayList<>();
        if (checkCallingOrSelfPermission(Manifest.permission.RECEIVE_SMS) != PackageManager.PERMISSION_GRANTED) {
            permissions.add(Manifest.permission.RECEIVE_SMS);
        }
        if (checkCallingOrSelfPermission(Manifest.permission.READ_SMS) != PackageManager.PERMISSION_GRANTED) {
            permissions.add(Manifest.permission.READ_SMS);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (checkCallingOrSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                permissions.add(Manifest.permission.POST_NOTIFICATIONS);
            }
        }

        if (!permissions.isEmpty()) {
            requestPermissions(permissions.toArray(new String[0]), PERMISSION_REQUEST_CODE);
        } else {
            SmsReaderForegroundService.start(this);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_REQUEST_CODE) {
            boolean smsGranted = false;
            for (int i = 0; i < permissions.length; i++) {
                if (Manifest.permission.RECEIVE_SMS.equals(permissions[i]) && grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                    smsGranted = true;
                    break;
                }
            }
            if (smsGranted) {
                SmsReaderForegroundService.start(this);
                Toast.makeText(this, "SMS Permission Granted. RJ World BD Payment Service is Active.", Toast.LENGTH_LONG).show();
                SyncManager.triggerSync(this);
            } else {
                SmsReaderForegroundService.stop(this, "SMS Permission unavailable");
                Toast.makeText(this, "SMS Permission Denied. Service cannot read payment SMS.", Toast.LENGTH_LONG).show();
            }
            updateStatusDisplay();
        }
    }
}

File: android-sms-reader/app/src/main/java/com/rjworldbd/smsreader/ui/MainActivity.kt
package com.rjworldbd.smsreader.ui

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import com.rjworldbd.smsreader.data.local.PaymentDatabase
import com.rjworldbd.smsreader.service.SmsReaderForegroundService
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

/**
 * Main Activity for RJ World BD SMS Reader.
 * Displays real-time metrics:
 * 1. Total Transactions
 * 2. Pending
 * 3. Synced
 * 4. Verified
 * 5. Rejected
 * 
 * Manages Foreground Service & SMS Permissions.
 */
class MainActivity : AppCompatActivity() {

    private val permissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        val smsGranted = permissions[Manifest.permission.RECEIVE_SMS] == true ||
                permissions[Manifest.permission.READ_SMS] == true

        if (smsGranted) {
            SmsReaderForegroundService.start(this)
            Toast.makeText(this, "SMS Permission Granted. RJ World BD Payment Service is Active.", Toast.LENGTH_LONG).show()
        } else {
            SmsReaderForegroundService.stop(this, "SMS Permission unavailable")
            Toast.makeText(this, "SMS Permission Denied. Service cannot read payment SMS.", Toast.LENGTH_LONG).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(com.rjworldbd.smsreader.R.layout.activity_main)

        // Check and request runtime permissions (SMS + Post Notifications on Android 13+)
        checkAndRequestPermissions()

        // Observe real-time database counters
        observeMetrics()
    }

    private fun checkAndRequestPermissions() {
        val permissions = mutableListOf(
            Manifest.permission.RECEIVE_SMS,
            Manifest.permission.READ_SMS
        )

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            permissions.add(Manifest.permission.POST_NOTIFICATIONS)
        }

        val missing = permissions.filter {
            ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED
        }

        if (missing.isNotEmpty()) {
            permissionLauncher.launch(missing.toTypedArray())
        } else {
            SmsReaderForegroundService.start(this)
        }
    }

    private fun observeMetrics() {
        val dao = PaymentDatabase.getInstance(this).paymentDao()

        lifecycleScope.launch {
            dao.getTotalCountFlow().collectLatest { count ->
                // Real-time total transactions count
            }
        }

        lifecycleScope.launch {
            dao.getPendingCountFlow().collectLatest { count ->
                // Real-time pending count
                SmsReaderForegroundService.updateCount(this@MainActivity, count)
            }
        }

        lifecycleScope.launch {
            dao.getSyncedCountFlow().collectLatest { count ->
                // Real-time synced count
            }
        }

        lifecycleScope.launch {
            dao.getVerifiedCountFlow().collectLatest { count ->
                // Real-time verified count
            }
        }

        lifecycleScope.launch {
            dao.getRejectedCountFlow().collectLatest { count ->
                // Real-time rejected count
            }
        }
    }
}
